### PR TITLE
Use circular buffer in LBFGS implementation.

### PR DESCRIPTION
**Pull request by mblondel**
Following @shoyer's idea, I implemented a circular buffer for maintaining the histories, thus avoiding memory copies. All tests are passing. I hope I didn't make any mistake!

### CODE REVIEW

**Comment by shoyer** [comment]
Nice! Have you tried running any micro-benchmarks to compare performance?

**Comment by soraros** [comment]
Maybe something like this is slightly cleaner, as the circular indexing is dealt with in a single place.
```python
def inv_hessian_product_leaf(v: jnp.ndarray,
                             s_history: jnp.ndarray,
                             y_history: jnp.ndarray,
                             rho_history: jnp.ndarray,
                             gamma: float = 1.0,
                             start: int = 0):
  history_size = len(s_history)
  I = (start - 1 - jnp.arange(history_size)) % history_size

  def body_right(r, i):
    alpha = rho_history[i] * jnp.vdot(s_history[i], r)
    r = r - alpha * y_history[i]
    return r, alpha

  r, alpha = lax.scan(body_right, v, I)

  r = r * gamma

  def body_left(r, tup):
    i, alpha = tup
    beta = rho_history[i] * jnp.vdot(y_history[i], r)
    r = r + s_history[i] * (alpha - beta)
    return r, beta

  r, beta = lax.scan(body_left, r, (I, alpha), reverse=True)
  return r
```

**Comment by mblondel** [comment]
@soraros I agree your version is cleaner. I switched to it, thanks a lot!

@shoyer On toy problems, the new version seems slightly faster. BTW, I plan to do a proper benchmark against SciPy's LBFGS in a separate PR.

File: jaxopt/_src/lbfgs.py [hunk 1]
```diff
@@ -38,39 +38,33 @@
...
 
-  # Compute right part.
-  q, alpha = jax.lax.scan(ihp_body_right,
-                          v,
-                          (s_history, y_history, rho_history),
-                          reverse=True)
+  def body_right(r, i):
+    alpha = rho_history[i] * jnp.vdot(s_history[i], r)
+    r = r - alpha * y_history[i]
+    return r, alpha
```
Comment (shoyer) [inline jaxopt/_src/lbfgs.py:55]: I guess you didn't need to handle pytrees in these helper functions after all?

**Review by shoyer — APPROVED** [review]

File: jaxopt/_src/lbfgs.py (hunk 1 above)
Comment (mblondel) [reply jaxopt/_src/lbfgs.py:55]: I'm not sure why but in the previous version of the code, I was getting errors if I didn't use tree utils, even though `inv_hessian_product_leaf` is supposed to only receive tree leaves... With the new version of the code, all tests passed directly with `jnp` functions so I went with that. Shall I use the tree utils instead?

Comment (shoyer) [reply jaxopt/_src/lbfgs.py:55]: You have test coverage, so if things work with the new code, that's fine by me!

**Pull request merged**